src/testing/AddTestToIdentifierPage.test.tsx

```diff
@@ -0,0 +1,75 @@
+import React from 'react';
+import { waitFor, fireEvent } from '@testing-library/react';
+import nock from 'nock';
+
+import { useAuthentication } from '../authentication/context';
+import { renderWrapped } from '../testHelpers';
+import { AddTestToIdentifierPage } from '.';
+
+jest.mock('../authentication/context');
+
+describe(AddTestToIdentifierPage, () => {
+  it('creates a user for the authentication method from config and the filled identifier', async () => {
```

Comment:
Any way to break this test up into smaller, more documented pieces?  It's quite hard to follow at the moment.

Reply:
Yup, will try to improve without creating too much repetition.

Do you feel comments and/or well named helper functions would help?

Reply:
I like to use BDD, pseudocode:
```
describe('when initialised', () => {
  beforeEach(() => {
    mount(...);
  });
  it('should show stuff ', () => { ... }

  describe('when text is entered the user clicks submit', () => {
    let promise;
    beforeEach(() => {
      button.click();
      promise = new Promise()
      mock(api.method).returnValue(promise)
    });
    it('should make a request to the xxx API', () => { ... });

    describe('when the request fails', () => {
      beforeEach(() => {
        promise.reject(...);
      });
      it('should show an error', () => { ...});
    });

    describe('when the request succeed', () => {
      beforeEach(() => {
        promise.resolve(...);
      });
      it('should ...', () => { ...});
    });
  });
});
```
A bit more verbose, but helps me lay stuff out in a readable way.  Don't feel obliged to follow that.

Reply:
I've followed BDD as well, but for React Testing Library's integration tests, I've started writing longer isolated tests and have found them to be more maintainable thanks to the immutability and isolation, especially with hooks and async operations.

Although slightly exaggerated, I agree with the main points from Kent C. Dodds:
https://kentcdodds.com/blog/write-fewer-longer-tests
https://kentcdodds.com/blog/avoid-nesting-when-youre-testing

That being said, a test definitely shouldn't lose its readability, so I'll figure out a way to improve 👍

Reply:
✅ Used `beforeEach` for the setup (mocking the token), added a `fillInput` helper function, and split apart testing the autofocus. When I went further, I started sacrificing readability, flow integration confidence, and maintainability.